src/Compiler/Driver/CompilerImports.fs

```diff
@@ -54,27 +55,57 @@ let (++) x s = x @ [ s ]
 
 let IsSignatureDataResource (r: ILResource) =
     r.Name.StartsWithOrdinal FSharpSignatureDataResourceName
+    || r.Name.StartsWithOrdinal FSharpSignatureCompressDataResourceName
     || r.Name.StartsWithOrdinal FSharpSignatureDataResourceName2
 
 let IsOptimizationDataResource (r: ILResource) =
     r.Name.StartsWithOrdinal FSharpOptimizationDataResourceName
+    || r.Name.StartsWithOrdinal FSharpOptimizationCompressDataResourceName
     || r.Name.StartsWithOrdinal FSharpOptimizationDataResourceName2
 
-let GetSignatureDataResourceName (r: ILResource) =
-    if r.Name.StartsWithOrdinal FSharpSignatureDataResourceName then
-        String.dropPrefix r.Name FSharpSignatureDataResourceName
-    elif r.Name.StartsWithOrdinal FSharpSignatureDataResourceName2 then
-        String.dropPrefix r.Name FSharpSignatureDataResourceName2
+let decompressResource (r: ILResource) =
+    use raw = r.GetBytes().AsStream()
+    use decompressed = new MemoryStream()
+    use deflator = new DeflateStream(raw, CompressionMode.Decompress)
+    deflator.CopyTo decompressed
+    deflator.Close()
+    ByteStorage.FromByteArray(decompressed.ToArray()).GetByteMemory()
+
+let GetResourceNameAndSignatureDataFunc (r: ILResource) =
+    let getResourceTypeAndName (r: ILResource) =
+        if r.Name.StartsWithOrdinal FSharpSignatureDataResourceName then
+            FSharpSignatureDataResourceName, String.dropPrefix r.Name FSharpSignatureDataResourceName
+        elif r.Name.StartsWithOrdinal FSharpSignatureCompressDataResourceName then
+            FSharpSignatureCompressDataResourceName, String.dropPrefix r.Name FSharpSignatureCompressDataResourceName
+        elif r.Name.StartsWithOrdinal FSharpSignatureDataResourceName2 then
+            FSharpSignatureDataResourceName2, String.dropPrefix r.Name FSharpSignatureDataResourceName2
+        else
+            failwith "GetSignatureDataResourceName"
+
+    let resourceType, ccuName = getResourceTypeAndName r
+
+    if resourceType = FSharpSignatureCompressDataResourceName then
+        ccuName, (fun () -> decompressResource (r))
     else
-        failwith "GetSignatureDataResourceName"
+        ccuName, (fun () -> r.GetBytes())
+
+let GetResourceNameAndOptimizationDataFunc (r: ILResource) =
+    let getResourceTypeAndName (r: ILResource) =
```

Comment:
This is a needless inner function - the `r` on the inside is the same as the `r` on the outside.

Just use this:

```fsharp
let GetResourceNameAndOptimizationDataFunc (r: ILResource) =
    let resourceType, ccuName =
        if r.Name.StartsWithOrdinal FSharpOptimizationDataResourceName then
            FSharpOptimizationDataResourceName, String.dropPrefix r.Name FSharpOptimizationDataResourceName
        elif r.Name.StartsWithOrdinal FSharpOptimizationCompressDataResourceName then
            FSharpOptimizationCompressDataResourceName, String.dropPrefix r.Name FSharpOptimizationCompressDataResourceName
        elif r.Name.StartsWithOrdinal FSharpOptimizationDataResourceName2 then
            FSharpOptimizationDataResourceName2, String.dropPrefix r.Name FSharpOptimizationDataResourceName2
        else
            failwith "GetOptimizationDataResourceName"
    if resourceType = FSharpOptimizationCompressDataResourceName then
        ccuName, (fun () -> decompressResource (r))
    else
        ccuName, (fun () -> r.GetBytes())
```

Reply:
done